proc/threads.c

```diff
@@ -467,49 +467,70 @@ static void thread_destroy(thread_t *t)
 	process_t *process;
 	spinlock_ctx_t sc;
 
-	perf_end(t);
+	proc_lockSet(&threads_common.lock);
 
+	lib_rbRemove(&threads_common.id, &t->idlinkage);
+
+	proc_lockClear(&threads_common.lock);
+
+	perf_end(t);
+	hal_spinlockDestroy(&t->spinlock);
 	vm_kfree(t->kstack);
 
 	if ((process = t->process) != NULL) {
 		hal_spinlockSet(&threads_common.spinlock, &sc);
+
 		LIST_REMOVE_EX(&process->threads, t, procnext, procprev);
 		LIST_ADD_EX(&process->ghosts, t, procnext, procprev);
 		_proc_threadWakeup(&process->reaper);
-		hal_spinlockClear(&threads_common.spinlock, &sc);
 
+		hal_spinlockClear(&threads_common.spinlock, &sc);
 		proc_put(process);
 	}
-	else
+	else {
 		vm_kfree(t);
+	}
 }
 
 
 thread_t *threads_findThread(int tid)
 {
-	thread_t *r, t;
-	t.id = tid;
+	thread_t *thread, t;
+	spinlock_ctx_t sc;
 
+	t.id = tid;
 	proc_lockSet(&threads_common.lock);
-	if ((r = lib_treeof(thread_t, idlinkage, lib_rbFind(&threads_common.id, &t.idlinkage))) != NULL)
-		r->refs++;
+
+	thread = lib_treeof(thread_t, idlinkage, lib_rbFind(&threads_common.id, &t.idlinkage));
+
 	proc_lockClear(&threads_common.lock);
 
-	return r;
+	if (thread != NULL) {
+		hal_spinlockSet(&thread->spinlock, &sc);
+
+		thread->refs++;
+
+		hal_spinlockClear(&thread->spinlock, &sc);
+	}
```

Comment:
Can we move this under threads_common.lock? There is small window (between lockClear and spinlockSet) during which thread is not protected from being destoyed

Reply:
Done. I've also added a check for soon to be destroyed thread in `threads_findThread()`.